{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3689923c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "175bce32",
   "metadata": {},
   "outputs": [],
   "source": [
    "DNA_AA_map = {\"TTT\":\"F\", \"TTC\":\"F\", \"TTA\":\"L\", \"TTG\":\"L\",\n",
    "              \"TCT\":\"S\", \"TCC\":\"S\", \"TCA\":\"S\", \"TCG\":\"S\",\n",
    "              \"TAT\":\"Y\", \"TAC\":\"Y\", \"TAA\":\".\", \"TAG\":\".\",\n",
    "              \"TGT\":\"C\", \"TGC\":\"C\", \"TGA\":\".\", \"TGG\":\"W\",\n",
    "              \"CTT\":\"L\", \"CTC\":\"L\", \"CTA\":\"L\", \"CTG\":\"L\",\n",
    "              \"CCT\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n",
    "              \"CAT\":\"H\", \"CAC\":\"H\", \"CAA\":\"Q\", \"CAG\":\"Q\",\n",
    "              \"CGT\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\",\n",
    "              \"ATT\":\"I\", \"ATC\":\"I\", \"ATA\":\"I\", \"ATG\":\"M\",\n",
    "              \"ACT\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n",
    "              \"AAT\":\"N\", \"AAC\":\"N\", \"AAA\":\"K\", \"AAG\":\"K\",\n",
    "              \"AGT\":\"S\", \"AGC\":\"S\", \"AGA\":\"R\", \"AGG\":\"R\",\n",
    "              \"GTT\":\"V\", \"GTC\":\"V\", \"GTA\":\"V\", \"GTG\":\"V\",\n",
    "              \"GCT\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\",\n",
    "              \"GAT\":\"D\", \"GAC\":\"D\", \"GAA\":\"E\", \"GAG\":\"E\",\n",
    "              \"GGT\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\", }\n",
    "\n",
    "base_editing_key = {\"CBE\": [\"C\", \"T\"], \n",
    "                      \"ABE\": [\"A\", \"G\"]\n",
    "                     }\n",
    "\n",
    "bases = 'ACGT'\n",
    "complements = {'A':'T', 'T':'A', 'G':'C', 'C':'G', \n",
    "                   'a':'t', 't':'a', 'g':'c', 'c':'g'}\n",
    "cas_key = {'Sp': 'NGG', 'SpG': 'NGN', 'SpRY': 'NNN'}\n",
    "\n",
    "def DNA_to_AA(seq): \n",
    "    aa_seq = ''\n",
    "    for i in range(len(seq)//3): \n",
    "        codon = seq[(i*3):(i*3)+3].replace(\"T\", \"U\")\n",
    "        aa_seq += DNA_AA_map[codon]\n",
    "    return aa_seq\n",
    "\n",
    "def rev_complement(complements, seq): \n",
    "    compl = ''\n",
    "    for i in range(len(seq)): \n",
    "        compl += complements[seq[i]]\n",
    "    return compl[::-1]\n",
    "\n",
    "def complement(complements, seq): \n",
    "    compl = ''\n",
    "    for i in range(len(seq)): \n",
    "        compl += complements[seq[i]]\n",
    "    return compl\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a7c0a99b",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Gene_BaseEditing(): \n",
    "    \n",
    "    def __init__(self, filepath, output_dir=''): \n",
    "        f = open(filepath, \"r\")\n",
    "        self.file_content = f.read()\n",
    "        self.exons_extra, self.exons = self.parse_exons()\n",
    "    \n",
    "    # function to read in a .fasta file as its exons, separating into just exon and exon +- 20 bps\n",
    "    # outputs: list of exons, list of exons +- 20 bps\n",
    "    def parse_exons(self): \n",
    "        exons_extra = []\n",
    "        i = -1\n",
    "        for line in self.file_content.split('\\n'): \n",
    "            if len(line) > 0 and line[0] == '>': \n",
    "                exons_extra.append('')\n",
    "                i += 1\n",
    "            else: \n",
    "                exons_extra[i] += line\n",
    "        exons = []\n",
    "        for exon in exons_extra: \n",
    "            exons.append(''.join([base for base in exon if base.isupper()]))\n",
    "        return exons_extra, exons\n",
    "    \n",
    "    # generates all fwd and rev potential guides and their metadata\n",
    "    # while this is computationally naive, genes are usually not that long\n",
    "    # pseudo outputs: guide sequence, the index of the first bp, the frame of the first bp, the exon # of guide\n",
    "    def find_all_guides(self, n): \n",
    "        self.fwd_guides = []\n",
    "        prev_frame = 0\n",
    "        prev_ind = 0\n",
    "        for e, exon_extra in enumerate(self.exons_extra): \n",
    "            for i in range(len(exon_extra)-22): \n",
    "                frame = (i+prev_frame-20)%3\n",
    "                ind = i-20+prev_ind\n",
    "                self.fwd_guides.append([exon_extra[i:i+23], frame, ind, e])\n",
    "            prev_frame = (prev_frame+len(exon_extra)-40)%3\n",
    "            prev_ind += len(exon_extra)-40\n",
    "            \n",
    "        self.rev_guides = [[rev_complement(complements, \n",
    "                                           g[0])] + [(g[1]+1)%3] + [g[2]+22] + [g[3]] for g in self.fwd_guides]\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7e4df30d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_gRNAs(gene_object, mode, cas_type, target_codons=[], window=[4,8], PAM=None): \n",
    "    # mode: can be CBE or ABE, or just a list of 2 base edits ex [\"C\", \"T\"]\n",
    "    # cas_type: Sp, SpG, SpRY\n",
    "    # target_codons: list of codons that we want to make with our base edit\n",
    "    # window: 4th to 8th bases inclusive by default, can be changed\n",
    "    # PAM: optional field to input a custom PAM\n",
    "    # Returns: a df of exon #, guides (23 bps), target (20 bps), fwd or rev, codon #, edit made\n",
    "    \n",
    "    # process mode\n",
    "    if len(mode) == 3: \n",
    "        mode = base_editing_key[mode]\n",
    "    assert len(mode) == 2\n",
    "    \n",
    "    # process PAM\n",
    "    if PAM is None: \n",
    "        PAM = cas_key[cas_type]\n",
    "    PAM_regex = process_PAM(PAM)\n",
    "    \n",
    "    # filter for PAM and contains editable base in window\n",
    "    fwd_results = [g.copy() for g in gene_object.fwd_guides if PAM_regex.match(g[0][-len(PAM):]) and \n",
    "                                                    mode[0] in g[0][window[0]-1:window[1]]]\n",
    "    for g in fwd_results: \n",
    "        # mutates all residues according to the mode\n",
    "        original = g[0][1:12]\n",
    "        mutated = g[0][1:window[0]-1] + g[0][window[0]-1:window[1]].replace(mode[0], \n",
    "                                                                            mode[1]) + g[0][window[1]:window[1]+4]\n",
    "        assert(len(original)==len(mutated))\n",
    "        # compares the residues to find which amino acids were altered and catalogs them\n",
    "        start = (-1*(g[1]-1))+1\n",
    "        original, mutated = original[start:start+9], mutated[start:start+9]\n",
    "        original_aa, mutated_aa = '', ''\n",
    "        edit, edit_ind = [], []\n",
    "        for i in range(3): \n",
    "            if not original[i*3:(i+1)*3].isupper(): \n",
    "                continue\n",
    "            original_aa += DNA_AA_map[original[i*3:(i+1)*3]]\n",
    "            mutated_aa += DNA_AA_map[mutated[i*3:(i+1)*3]]\n",
    "            if original_aa[-1] != mutated_aa[-1]: \n",
    "                edit.append(original_aa[-1] + \">\" + mutated_aa[-1])\n",
    "                edit_ind.append(int((g[2]+1+start+(i*3))/3))\n",
    "                \n",
    "        if len(edit) == 0: \n",
    "            edit.append('No Change')\n",
    "        # append all information to dataframe\n",
    "        g.append(original_aa)\n",
    "        g.append(mutated_aa)\n",
    "        g.append(edit)\n",
    "        g.append(edit_ind)\n",
    "        g.append('fwd')\n",
    "        assert(len(g)) == 9\n",
    "        \n",
    "    # filter for PAM and contains editable base in window\n",
    "    rev_results = [g.copy() for g in gene_object.rev_guides if PAM_regex.match(g[0][-len(PAM):]) and \n",
    "                                                    mode[0] in g[0][window[0]-1:window[1]]]\n",
    "    for g in rev_results: \n",
    "        # mutates all residues according to the mode\n",
    "        original = g[0][1:12]\n",
    "        mutated = g[0][1:window[0]-1] + g[0][window[0]-1:window[1]].replace(mode[0], \n",
    "                                                                            mode[1]) + g[0][window[1]:window[1]+4]\n",
    "        assert(len(original)==len(mutated))\n",
    "        # compares the residues to find which amino acids were altered and catalogs them\n",
    "        original = rev_complement(complements, original[g[1]:g[1]+9])\n",
    "        mutated = rev_complement(complements, mutated[g[1]:g[1]+9])\n",
    "        original_aa, mutated_aa = '', ''\n",
    "        edit, edit_ind = [], []\n",
    "        for i in range(3): \n",
    "            if not original[i*3:(i+1)*3].isupper(): \n",
    "                continue\n",
    "            original_aa += DNA_AA_map[original[i*3:(i+1)*3]]\n",
    "            mutated_aa += DNA_AA_map[mutated[i*3:(i+1)*3]]\n",
    "            if original_aa[-1] != mutated_aa[-1]: \n",
    "                edit.append(original_aa[-1] + \">\" + mutated_aa[-1])\n",
    "                edit_ind.append(int((g[2]-5+(i*3))/3))\n",
    "                \n",
    "        if len(edit) == 0: \n",
    "            edit.append('No Change')\n",
    "        # append all information to dataframe\n",
    "        g.append(original_aa)\n",
    "        g.append(mutated_aa)\n",
    "        g.append(edit)\n",
    "        g.append(edit_ind)\n",
    "        g.append('rev')\n",
    "        assert(len(g)) == 9\n",
    "        \n",
    "    results = fwd_results + rev_results\n",
    "    return pd.DataFrame(results)\n",
    "\n",
    "# function to change a PAM sequence into a regex sequence\n",
    "def process_PAM(PAM): \n",
    "    PAM = PAM.replace(\"G\", \"[gG]{1}\")\n",
    "    PAM = PAM.replace(\"C\", \"[cC]{1}\")\n",
    "    PAM = PAM.replace(\"T\", \"[tT]{1}\")\n",
    "    PAM = PAM.replace(\"A\", \"[aA]{1}\")\n",
    "    PAM = PAM.replace(\"N\", \"[acgtACGT]{1}\")\n",
    "    return re.compile(\"({})\".format(PAM))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9f67ea0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAGTGCAGTTAGGGCTGGGAAGG</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>VQL</td>\n",
       "      <td>V.L</td>\n",
       "      <td>[Q&gt;.]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>fwd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGTGCAGTTAGGGCTGGGAAGGG</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>VQL</td>\n",
       "      <td>V.L</td>\n",
       "      <td>[Q&gt;.]</td>\n",
       "      <td>[3]</td>\n",
       "      <td>fwd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CCGCCGTCCAAGACCTACCGAGG</td>\n",
       "      <td>0</td>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "      <td>PSK</td>\n",
       "      <td>LFK</td>\n",
       "      <td>[P&gt;L, S&gt;F]</td>\n",
       "      <td>[14, 15]</td>\n",
       "      <td>fwd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCGCGAAGTGATCCAGAACCCGG</td>\n",
       "      <td>2</td>\n",
       "      <td>89</td>\n",
       "      <td>0</td>\n",
       "      <td>REV</td>\n",
       "      <td>REV</td>\n",
       "      <td>[No Change]</td>\n",
       "      <td>[]</td>\n",
       "      <td>fwd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TGATCCAGAACCCGGGCCCCAGG</td>\n",
       "      <td>1</td>\n",
       "      <td>97</td>\n",
       "      <td>0</td>\n",
       "      <td>IQN</td>\n",
       "      <td>I.N</td>\n",
       "      <td>[Q&gt;.]</td>\n",
       "      <td>[34]</td>\n",
       "      <td>fwd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>374</th>\n",
       "      <td>ATGATCTCTGCCATCATTTCCGG</td>\n",
       "      <td>1</td>\n",
       "      <td>2698</td>\n",
       "      <td>7</td>\n",
       "      <td>AEI</td>\n",
       "      <td>AKI</td>\n",
       "      <td>[E&gt;K]</td>\n",
       "      <td>[898]</td>\n",
       "      <td>rev</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>ACTTTCCCAGAAAGGATCTTGGG</td>\n",
       "      <td>1</td>\n",
       "      <td>2734</td>\n",
       "      <td>7</td>\n",
       "      <td>SGK</td>\n",
       "      <td>SKK</td>\n",
       "      <td>[G&gt;K]</td>\n",
       "      <td>[910]</td>\n",
       "      <td>rev</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>GACTTTCCCAGAAAGGATCTTGG</td>\n",
       "      <td>2</td>\n",
       "      <td>2735</td>\n",
       "      <td>7</td>\n",
       "      <td>SGK</td>\n",
       "      <td>SEK</td>\n",
       "      <td>[G&gt;E]</td>\n",
       "      <td>[911]</td>\n",
       "      <td>rev</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>TGGGCTTGACTTTCCCAGAAAGG</td>\n",
       "      <td>0</td>\n",
       "      <td>2742</td>\n",
       "      <td>7</td>\n",
       "      <td>VKP</td>\n",
       "      <td>VKP</td>\n",
       "      <td>[No Change]</td>\n",
       "      <td>[]</td>\n",
       "      <td>rev</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>TCACTGGGTGTGGAAATAGATGG</td>\n",
       "      <td>2</td>\n",
       "      <td>2762</td>\n",
       "      <td>7</td>\n",
       "      <td>HTQ</td>\n",
       "      <td>HTQ</td>\n",
       "      <td>[No Change]</td>\n",
       "      <td>[]</td>\n",
       "      <td>rev</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>379 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           0  1     2  3    4    5            6         7    8\n",
       "0    AAGTGCAGTTAGGGCTGGGAAGG  1     4  0  VQL  V.L        [Q>.]       [3]  fwd\n",
       "1    AGTGCAGTTAGGGCTGGGAAGGG  2     5  0  VQL  V.L        [Q>.]       [3]  fwd\n",
       "2    CCGCCGTCCAAGACCTACCGAGG  0    39  0  PSK  LFK   [P>L, S>F]  [14, 15]  fwd\n",
       "3    GCGCGAAGTGATCCAGAACCCGG  2    89  0  REV  REV  [No Change]        []  fwd\n",
       "4    TGATCCAGAACCCGGGCCCCAGG  1    97  0  IQN  I.N        [Q>.]      [34]  fwd\n",
       "..                       ... ..   ... ..  ...  ...          ...       ...  ...\n",
       "374  ATGATCTCTGCCATCATTTCCGG  1  2698  7  AEI  AKI        [E>K]     [898]  rev\n",
       "375  ACTTTCCCAGAAAGGATCTTGGG  1  2734  7  SGK  SKK        [G>K]     [910]  rev\n",
       "376  GACTTTCCCAGAAAGGATCTTGG  2  2735  7  SGK  SEK        [G>E]     [911]  rev\n",
       "377  TGGGCTTGACTTTCCCAGAAAGG  0  2742  7  VKP  VKP  [No Change]        []  rev\n",
       "378  TCACTGGGTGTGGAAATAGATGG  2  2762  7  HTQ  HTQ  [No Change]        []  rev\n",
       "\n",
       "[379 rows x 9 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1. generate a .fasta file of your gene using this protocol \"Creating CRISPR sgRNA Libraries on CRISPOR.pdf\"\n",
    "# 2. inputs\n",
    "#       filepath is your fasta file\n",
    "#       output_dir can be any folder\n",
    "#       n= length of your guide, 23 is standard\n",
    "#       mode (ABE, CBE, or a list of edits like [\"A\", \"C\"])\n",
    "#       cas_type (Sp, SpG)\n",
    "#       target_codons=[] no targets default but you can look for specific amino acid residues you want to generate\n",
    "#       window=[4,8] 4-8 inclusive is the default BE window but you can edit this\n",
    "#       PAM=None default but you can input custom PAM (ex NGNG), if you leave as None the PAM is decided by cas_type\n",
    "# 3. run the following code\n",
    "\n",
    "AR = Gene_BaseEditing(filepath=\"230408_AR_Input.fasta\", \n",
    "                      output_dir=\"results\"\n",
    "                     )\n",
    "AR.find_all_guides(n=23)\n",
    "find_gRNAs(gene_object= AR, \n",
    "           mode=        \"CBE\", \n",
    "           cas_type=    \"Sp\"\n",
    "          ) #.to_csv('230409_AR_BESpCBE_library.csv') # if you want to save this dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd863da3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lab",
   "language": "python",
   "name": "lab"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
