{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8054f2cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import itertools\n",
    "from itertools import permutations\n",
    "import re\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "3b3fe21a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# a CBE produces C>T on sense strand when fwd, G>A on sense strand if rev\n",
    "# a ABE produces A>G on sense strand when fwd, T>C on sense strand if rev\n",
    "\n",
    "# usable for PAMs: Sp NGG, SpG NGN, SpRY NNN\n",
    "# \n",
    "# \n",
    "\n",
    "class BE_gRNAs(): \n",
    "    \n",
    "    # references\n",
    "    bases = 'ACGT'\n",
    "    complements = {'A':'T', 'T':'A', 'G':'C', 'C':'G', \n",
    "                   'a':'t', 't':'a', 'g':'c', 'c':'g'}\n",
    "    cas_key = {'Sp': 'NGG', 'SpG': 'NGN', 'SpRY': 'NNN'}\n",
    "    \n",
    "    def __init__(self, be_type, editing_window, gene, end_goals, cas_type, exon_filename, exon_dir=''): \n",
    "        \n",
    "        # vars\n",
    "        self.be_type = be_type\n",
    "        self.gene = gene\n",
    "        self.end_goals = end_goals\n",
    "        self.PAM = self.cas_key[cas_type]\n",
    "        self.editing_window = editing_window\n",
    "        \n",
    "        # load file of exons\n",
    "        self.filename = os.path.join(exon_dir, exon_filename)\n",
    "        f = open(self.filename, \"r\")\n",
    "        self.file_content = f.read()\n",
    "        # list of exons and list of exons with +-20 bps of introns\n",
    "            # both are necessary bc +-20 bps needed bc gRNA can attach onto DNA that isnt expressed\n",
    "            # only +- 20 is needed bc gRNA is only 23 bps\n",
    "        self.exons_extra, self.exons = self.parse_exons()\n",
    "        \n",
    "        if len(self.end_goals) > 0: \n",
    "    \n",
    "            # find all sense and antisense codons that can be mutated\n",
    "            self.target_codons, self.target_codons_compls = self.generate_target_codons()\n",
    "            # find all indices with in frame codon, in range PAM\n",
    "            self.indices = self.find_all_indices()\n",
    "            # find all full guide RNAs\n",
    "            self.potential_gRNAs, self.sense_antisense, self.exon_num, self.targets = self.get_gRNAs()\n",
    "                    \n",
    "        else: \n",
    "            \n",
    "            # find all full guide RNAs without a target codon to change\n",
    "            self.potential_gRNAs, self.sense_antisense, self.exon_num, self.targets = self.get_all_gRNAs()\n",
    "            \n",
    "        # self.gene self.potential_gRNAs self.sense_antisense self.exon_num self.targets\n",
    "    \n",
    "    ###################################################################################################\n",
    "\n",
    "    # parse exons with and without intron ends from fasta file format\n",
    "    def parse_exons(self): \n",
    "        exons_extra = []\n",
    "        i = -1\n",
    "        for line in self.file_content.split('\\n'): \n",
    "            if len(line) > 0 and line[0] == '>': \n",
    "                exons_extra.append('')\n",
    "                i += 1\n",
    "            else: \n",
    "                exons_extra[i] += line\n",
    "\n",
    "        exons = []\n",
    "        for exon in exons_extra: \n",
    "            exons.append(''.join([base for base in exon if base.isupper()]))\n",
    "        return exons_extra, exons\n",
    "    \n",
    "    def rev_complement(self, seq): \n",
    "        compl = ''\n",
    "        for i in range(len(seq)): \n",
    "            compl += self.complements[seq[i]]\n",
    "        return compl[::-1]\n",
    "    \n",
    "    def complement(self, seq): \n",
    "        compl = ''\n",
    "        for i in range(len(seq)): \n",
    "            compl += self.complements[seq[i]]\n",
    "        return compl\n",
    "    \n",
    "    ###################################################################################################\n",
    "    \n",
    "    # generate target codons based on type of base editor, and what codon we want to generate\n",
    "    def generate_target_codons(self): \n",
    "        codons = []\n",
    "        anticodons = []\n",
    "        for goal in self.end_goals: \n",
    "            codons += self.base_edits(goal, 'sense')\n",
    "            anticodons += self.base_edits(goal, 'anti')\n",
    "        return list(set(codons)), list(set(anticodons))\n",
    "    \n",
    "    ### horrible\n",
    "    def base_edits(self, goal, mode): \n",
    "        if self.be_type == 'CBE': \n",
    "            if mode == 'sense': \n",
    "                return self.replace(goal, 'C', 'T')\n",
    "            elif mode == 'anti': \n",
    "                return self.replace(goal, 'G', 'A')\n",
    "        elif self.be_type == 'ABE':\n",
    "            if mode == 'sense': \n",
    "                return self.replace(goal, 'A', 'G')\n",
    "            elif mode == 'anti': \n",
    "                return self.replace(goal, 'T', 'C')\n",
    "        else: \n",
    "            print('BE type invalid')\n",
    "        \n",
    "    def replace(self, goal, x, y): \n",
    "        result = [goal[0:i] + x + goal[i+1:] for i in range(3) if goal[i]==y]\n",
    "        result += [goal[0:i] + x + x + goal[i+2:] for i in range(2) if goal[i:i+2]==y+y]\n",
    "        result += [x + goal[1] + x for i in range(1) if goal[0]==y and goal[2]==y]\n",
    "        result += [x+x+x for i in range(1) if goal==y+y+y]\n",
    "        return result\n",
    "        \n",
    "    ###################################################################################################\n",
    "    \n",
    "    def find_all_indices(self): \n",
    "        result = [[]] * len(self.exons_extra)\n",
    "        for i in range(len(self.exons)): \n",
    "            indices = []\n",
    "            prev_len = sum([len(exon) for exon in self.exons[:i]])\n",
    "            for codon in self.target_codons: \n",
    "                inds = [m.start() for m in re.finditer(codon, self.exons_extra[i])]\n",
    "                inds = [ind for ind in inds if ((ind+prev_len+1)%3)==0]\n",
    "                inds = [[ind, 'fwd', i+1] for ind in inds if self.has_PAM_CBE(i, ind, 'fwd')]\n",
    "                indices += inds\n",
    "            for codon_compl in self.target_codons_compls: \n",
    "                inds = [m.start() for m in re.finditer(codon_compl, self.exons_extra[i])]\n",
    "                inds = [ind for ind in inds if ((ind+prev_len+1)%3)==0]\n",
    "                inds = [[ind, 'rev', i+1] for ind in inds if self.has_PAM_CBE(i, ind, 'rev')]\n",
    "                indices += inds\n",
    "            indices.sort()\n",
    "            result[i] = indices\n",
    "        return result\n",
    "    \n",
    "    def has_PAM_CBE(self, i, ind, direction): \n",
    "        real_PAM = self.PAM.replace(\"N\", \"\")\n",
    "        if direction == 'fwd':\n",
    "            frame = self.exons_extra[i][ind:ind+20]\n",
    "            return real_PAM in frame[-6:]\n",
    "        elif direction == 'rev':\n",
    "            frame = self.exons_extra[i][ind-17:ind+3]\n",
    "            real_PAM = self.rev_complement(real_PAM)\n",
    "            return real_PAM in frame[:6]\n",
    "    \n",
    "    ###################################################################################################\n",
    "\n",
    "    def get_gRNAs(self): \n",
    "        potential_gRNAs = []\n",
    "        potential_target = []\n",
    "        sense_antisense = []\n",
    "        exon_num = []\n",
    "        real_PAM = self.PAM.replace(\"N\", \"\")\n",
    "\n",
    "        # make list of all guide RNAs possible\n",
    "        for i in range(len(self.exons)): \n",
    "            for j in range(len(self.indices[i])):\n",
    "                frame = self.find_frame(self.exons_extra[i], \n",
    "                                        self.indices[i][j][0],\n",
    "                                        self.indices[i][j][1])\n",
    "                for k in range(5): \n",
    "                    if frame[k+21:k+21+len(real_PAM)] == real_PAM: ###\n",
    "                        potential_gRNAs.append(frame[k:k+23])\n",
    "                        sense_antisense.append(self.indices[i][j][1])\n",
    "                        if self.indices[i][j][1] == 'fwd': \n",
    "                            potential_target.append(frame[k:k+23])\n",
    "                        elif self.indices[i][j][1] == 'rev':\n",
    "                            potential_target.append(self.rev_complement(frame[k:k+23]))                            \n",
    "                        exon_num.append(self.indices[i][j][2])\n",
    "        \n",
    "        return potential_gRNAs, sense_antisense, exon_num, potential_target\n",
    "        \n",
    "    def find_frame(self, exon, ind, direction): \n",
    "        if direction=='fwd': \n",
    "            return exon[ind-7:ind+20]\n",
    "        elif direction=='rev': \n",
    "            return self.rev_complement(exon[ind-17:ind+10])\n",
    "\n",
    "    ###################################################################################################\n",
    "    \n",
    "    def get_all_gRNAs(self): \n",
    "        potential_gRNAs = []\n",
    "        potential_target = []\n",
    "        sense_antisense = []\n",
    "        exon_num = []\n",
    "        real_PAM = self.PAM.replace(\"N\", \"\")\n",
    "        \n",
    "        # make list of all guide RNAs possible\n",
    "        for i in range(len(self.exons)): \n",
    "            exon_i = self.exons_extra[i]\n",
    "            for j in range(len(exon_i)): \n",
    "\n",
    "                # looking fwd\n",
    "                # NGG and NGN cases can be generalized since 21 starts at GG or GN                \n",
    "                frame = exon_i[j:j+23]\n",
    "                if frame[21:21+len(real_PAM)].upper() == real_PAM: \n",
    "                    sub_frame = frame[self.editing_window[0]-1:self.editing_window[1]]\n",
    "                    if (self.be_type == 'CBE' and 'C' in sub_frame) or (self.be_type == 'ABE' and 'A' in sub_frame): \n",
    "                        potential_gRNAs.append(frame)\n",
    "                        potential_target.append(frame)\n",
    "                        sense_antisense.append('fwd')\n",
    "                        exon_num.append(i+1)\n",
    "                \n",
    "                # looking rev\n",
    "                # since the opposite just involves looking at the opposite strand, same code but just rev_compl\n",
    "                revcompl_frame = self.rev_complement(frame)\n",
    "                if revcompl_frame[21:21+len(real_PAM)].upper() == real_PAM: \n",
    "                    sub_frame = revcompl_frame[self.editing_window[0]-1:self.editing_window[1]]\n",
    "                    if (self.be_type == 'CBE' and 'C' in sub_frame) or (self.be_type == 'ABE' and 'A' in sub_frame): \n",
    "                        potential_gRNAs.append(revcompl_frame)\n",
    "                        potential_target.append(self.rev_complement(revcompl_frame))\n",
    "                        sense_antisense.append('rev')\n",
    "                        exon_num.append(i+1)\n",
    "\n",
    "        return potential_gRNAs, sense_antisense, exon_num, potential_target\n",
    "\n",
    "    ###################################################################################################\n",
    "\n",
    "    def save_data(self, filename): \n",
    "        head = ['gene', 'guide RNA', 'target', 'direction', 'exon']\n",
    "        df = pd.DataFrame(zip([self.gene]*len(self.potential_gRNAs), \n",
    "                              self.potential_gRNAs, \n",
    "                              self.targets,\n",
    "                              self.sense_antisense, \n",
    "                              self.exon_num), \n",
    "                          columns=head)\n",
    "        df.to_csv(filename, index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69d4496a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "2b8c5668",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cas_type should be Sp, SpG, SpRY\n",
    "# base_editor_type should be ABE, CBE\n",
    "# stop_codons can be all stop codons, or any other codons you want to generate, \n",
    "#     or an empty list [] if you want a list of all possible guides\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "ea8410a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# example\n",
    "\n",
    "###\n",
    "base_editor_type = 'CBE'\n",
    "target_codons = ['TAG', 'TAA', 'TGA'] # to generate stop codons = ['TAG', 'TAA', 'TGA']\n",
    "cas_type = 'Sp'\n",
    "editing_window = (4, 8) # inclusive\n",
    "notes = 'stopCodons'\n",
    "###\n",
    "\n",
    "gene = 'ARexon8'\n",
    "exon_filename = '221210_AR_Exon8_Input.fasta'\n",
    "output_dir = 'results/'\n",
    "\n",
    "AR_exon8_BE = BE_gRNAs(base_editor_type, editing_window, gene, target_codons, cas_type, exon_filename)\n",
    "AR_exon8_BE.save_data(output_dir + gene + '_' + cas_type + base_editor_type + '_' + notes + '_gRNAs.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "1390d305",
   "metadata": {},
   "outputs": [],
   "source": [
    "# bugs in code\n",
    "# a stop codon turning into another stop codon should be excluded\n",
    "# for SpRY and a lesser extent SpG and Sp\n",
    "#    the same site is targeted by 5 gRNAs since SpRY has no PAM and moving down one base makes a new gRNA\n",
    "#    these should be somehow classified together as part of a subgroup\n",
    "# add a direction option\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "7979b6bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "136"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('ATTGCGAGAGAGCTGCATCAGTTCGCTTTTGACCTGCTAATCAAGTCACACATGGTGAGCGTGGACTTTCCGGAAATGATGGCAGAGATCATCTCTGTGCAAGTGCCCAAGATCCTTTCTGGGAAAGTCAAGCCCA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27e899bc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
